app/controllers/sessions_controller.rb

```diff
@@ -296,4 +313,22 @@ def maybe_skip_to_sign_up
     redirect_to signup_path if %w{signup student_signup}.include?(params[:go])
   end
 
+  def maybe_launch_with_trusted_params
+    return unless params[:go] == 'trusted_launch'
+
+    base_string = OAuth::Helper.normalize(
+      params.except(:controller, :action, :client_id, :signature)
+    )
+    secret_key = Doorkeeper::Application.find_by_uid!(params[:client_id]).secret
+    signature = OpenSSL::HMAC.hexdigest('sha1', secret_key, base_string)
+    if signature != params[:signature] ||
```

Comment:
being paranoid here... can `hexdigest` ever return `nil`?  If so we need a separate error for `params[:signature]` being `nil`